servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutDemandPublisher.java

```diff
@@ -131,6 +138,26 @@ private void startTimer() {
                 final Cancellable nextTimer = parent.timeoutExecutor.schedule(this::timerFires, parent.durationNs,
                         NANOSECONDS);
                 if (timerCancellableUpdater.compareAndSet(this, cancellable, nextTimer)) {
+                    assert cancellable == null;
+                    // We don't atomically manipulate the `demand` and set the timer so we can get the following race:
+                    // Thread1: `.onNext(..)` call decrements demand to 0 and enters the top of `startTimer()`, pauses.
+                    // Thread2: `.request(n > 0)`, increments demand from 0 -> n, enters `stopTimer(false)` and replaces
+                    // `null` with `null`.
+                    // Thread1: wakes and proceeds to successfully `startTimer()` despite demand being n > 0.
+                    // If we see demand > 0 here we know it is safe to stop the timer because this method is only called
+                    // on the Subscriber thread (no concurrency allowed), otherwise we let the timer stand.
+                    for (;;) {
+                        final long currDemand = demand;
+                        if (currDemand > 0) {
+                            nextTimer.cancel();
+                            // Try to reset the timerCancellableUpdater to the original state. If the CAS fails, no need
+                            // to loop because the only other state is LOCAL_IGNORE_CANCEL which is a terminal state.
+                            timerCancellableUpdater.compareAndSet(this, nextTimer, null);
+                            break;
+                        } else if (demandUpdater.compareAndSet(this, currDemand, currDemand)) {
+                            break;
+                        }
```

Comment:
`currDemand > 0` -> if we see this we know we can cancel the timer an bail (this is called on the Subscriber thread so there will be no decrements of the count).

 `currDemand == 0` -> if we see this, we need to confirm the value we read hasn't been changed by another thread (e.g. incremented due to more demand). it is possible the value we read on this thread is stale, and CAS will ensure it isn't stale.

`currDemand < 0` -> just in case the timer fired on another thread we break out of the loop

Reply:
I'll clarify/add comments.

Reply:
I'm still not seeing it: the nugget is the `if (demand > 0)` check and the `for` loop and `cas` still seem unnecessary. We read the latest value of `demand` by virtue of reading the volatile, and the race conditions with other threads incrementing the counter in the meantime seem intrinsic and can still happen right after the `cas`.

Do you have a concrete example that may make it more clear? It doesn't seem broken, just unnecessary, so I'm fine to let it go.